### PR TITLE
Update CNN encoder decoder logic

### PR DESCRIPTION
**Pull request by Prashanthsrn**
### 1. Network Architecture

```python
# Previous 
x = Conv2D(32, (3, 3), strides=(2, 2))(input_img)  
x = Conv2D(16, (3, 3), strides=(2, 2))(x)         

# New
x = Conv2D(32, (3, 3), activation='relu', padding='same')(input_img)
x = BatchNormalization()(x)
skip1 = x  # Store for skip connection

x = Conv2D(64, (3, 3), strides=2, activation='relu', padding='same')(x)
x = BatchNormalization()(x)
skip2 = x  # Store for skip connection
```

### 2. Added Skip Connections
Skip connections added

```python
# Decoder with skip connections
x = Conv2DTranspose(64, (3, 3), strides=2)(x)
x = Add()([x, skip2])  # Add features from encoder path

x = Conv2DTranspose(32, (3, 3), strides=2)(x)
x = Add()([x, skip1])  # Add features from encoder path
```

### 3. Training Parameters
```python
# Previous settings
self.epochs = 50
self.batch_size = 5

# Improved settings
self.epochs = 100        
self.batch_size = 16     
```

### 4. Added BatchNormalization

```python
x = Conv2D(64, (3, 3), activation='relu', padding='same')(x)
x = BatchNormalization()(x)  
```

### 5. Training Improvements
Added callbacks for training control:

```python
callbacks = [
    EarlyStopping(
        monitor='loss',
        patience=10,
        restore_best_weights=True
    ),
    ReduceLROnPlateau(
        monitor='loss',
        factor=0.5,
        patience=5,
        min_lr=1e-6
    )
]

model.fit(
    x_train, y_train,
    epochs=100,
    batch_size=16,
    callbacks=callbacks,
    validation_split=0.2
)
```

### CODE REVIEW

**Review by yogeshhk — APPROVED** [review]
Good work
